spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryJsonDataWriter.java

```diff
@@ -0,0 +1,148 @@
+
+    CreateWriteStreamRequest createWriteStreamRequest =
+        CreateWriteStreamRequest.newBuilder()
+            .setParent(parentTable.toString())
+            .setWriteStream(stream)
+            .build();
+    WriteStream writeStream = client.createWriteStream(createWriteStreamRequest);
+
+    // Use the JSON stream writer to send records in JSON format.
+    // For more information about JsonStreamWriter, see:
+    // https://googleapis.dev/java/google-cloud-bigquerystorage/latest/com/google/cloud/bigquery/storage/v1beta2/JsonStreamWriter.html
+    streamWriter =
+        JsonStreamWriter.newBuilder(writeStream.getName(), writeStream.getTableSchema()).build();
+  }
+
+  public void append(JSONArray data, long offset)
+      throws DescriptorValidationException, IOException, ExecutionException {
+    synchronized (this.lock) {
+      // If earlier appends have failed, we need to reset before continuing.
+      if (this.error != null) {
+        throw this.error;
+      }
+    }
+    // Append asynchronously for increased throughput.
+    ApiFuture<AppendRowsResponse> future = streamWriter.append(data, offset);
+    ApiFutures.addCallback(
+        future, new AppendCompleteCallback(this), MoreExecutors.directExecutor());
+    // Increase the count of in-flight requests.
+    inflightRequestCount.register();
+  }
+
+  public void cleanup(BigQueryWriteClient client) {
+    // Wait for all in-flight requests to complete.
+    inflightRequestCount.arriveAndAwaitAdvance();
+
+    // Close the connection to the server.
+    streamWriter.close();
+
+    // Verify that no error occurred in the stream.
+    synchronized (this.lock) {
+      if (this.error != null) {
+        throw this.error;
+      }
+    }
+
+    // Finalize the stream.
+    FinalizeWriteStreamResponse finalizeResponse =
+        client.finalizeWriteStream(streamWriter.getStreamName());
+    logger.log(Level.FINE, "\nRows written: " + finalizeResponse.getRowCount());
+  }
+
+  public String getStreamName() {
+    return streamWriter.getStreamName();
+  }
+
+  static class AppendCompleteCallback implements ApiFutureCallback<AppendRowsResponse> {
+    private final Logger logger = Logger.getLogger(this.getClass().getName());
+    private final BigQueryJsonDataWriter parent;
+
+    public AppendCompleteCallback(BigQueryJsonDataWriter parent) {
+      this.parent = parent;
+    }
+
+    public void onSuccess(AppendRowsResponse response) {
+      logger.log(
+          Level.FINE,
+          String.format("\nAppend %d success", response.getAppendResult().getOffset().getValue()));
+      done();
+    }
+
+    public void onFailure(Throwable throwable) {
+      synchronized (this.parent.lock) {
+        if (this.parent.error == null) {
+          StorageException storageException = Exceptions.toStorageException(throwable);
+          this.parent.error =
+              (storageException != null) ? storageException : new RuntimeException(throwable);
```

Comment:
I'm coming in to this a little late in the game, but has there been discussion around the error handling design for this class?

Wrapping this Throwable in a RuntimeException and only saving the first one seems like it might create some limitations for the user of this class to get the full context of the problem -- particularly if they're performing a `catch` on a particular exception that may now be hidden one level down within the wrapper.

Should the caller of the BigQueryJsonDataWriter provide an onFailure callback? Or perhaps the BigQueryJsonDataWriter.append method should take the AppendCompleteCallback as a parameter? (In both cases: the goal being to delegate the error handling decisions one level higher.)

I'm not sure. Just for consideration.

Reply:
Makes sense and thanks for pointing this. I think we will need to improve the error handling design here. Say a user is writing data in bulk, like billions of records and say append failed for a few thousand. Then we should be having a mechanism to pass this message back to the user without interrupting the whole operation (as we might be still good as far as analytics is concerned).This same applies while committing the records. Let me take it back to the drawing board and address this in a followup PR.  Let me have a word with @elefeint  as well on this.

Reply:
captured as an AI @ go/bq-write-api-sc-gcp-integration